Lookup parts/partitions in replica's own detached/ folder before downloading them from other replicas

myrrc opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Backward Incompatible Change


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now replicas that are processing the `ALTER TABLE ATTACH PART[ITION]` command search in their `detached/` folders before fetching the data from other replicas. As an implementation detail, a new command `ATTACH_PART` is introduced in the replicated log. Parts are searched and compared by their checksums.

Note:
- `ATTACH PART[ITION]` queries may not work during cluster upgrade.
- It's not possible to rollback to older ClickHouse version after executing `ALTER ... ATTACH` query in new version as the old servers would fail to pass the `ATTACH_PART` entry in the replicated log.

myrrc committed:
marked places to edit

myrrc committed:
Started adding the new "ATTACH_PART" command into the replicated log

The original ticket idea was to search for the possibly available data
into the /detached folders for the GET_PART command, but
@tavplubix pointed out this would be quite expensive for an every
fetch.

So a new command is going to be introduced, ATTACH_PART, which will
cover ALTER TABLE ATTACH PART and only for which the search will start.

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/attach-partition-local

myrrc committed:
Trying to re-write the solution by adding new command type

ATTACH_PART into the replicated log.

The LogEntry now also has the pre-calculated part checksum for this
entry type, which is later used while searching in the detached/ folder

myrrc committed:
Add some code for the checksum pre-calculation in ...BlockOutputStream

Added the comment explaining the double-get for the zookeeper header for
the part.

myrrc committed:
Add test stub

myrrc committed:
Correct checksums pre-calculation for ATTACH_PART in ...OutputStream

myrrc committed:
Replaced the part lookup algo to "by hash only", comments on test stub

myrrc committed:
Multiple small hotfixes

Small fixes

Some fix for old bug

Another old code fix

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/attach-partition-local

myrrc committed:
Fixed the UB with maybe-nullptr member access

myrrc committed:
Added part_checksum to Replicated...Entry serialization.

myrrc committed:
Trying to figure out why iteration doesn't work

myrrc committed:
Adding the part when found

myrrc committed:
Exit fixes

myrrc committed:
Finished the test draft for ATTACH PARTITION,

Extracted the part data corruption function into the helper.

myrrc committed:
Extended the test to check both the ALTER PARTITION and PART

Added some notes about the SYSTEM SYNC REPLICA and ALTER ... DROP /
ATTACH.

myrrc committed:
Fixing the tests

myrrc committed:
Added the PartitionManager to check that replica 1 attaches the local
data

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/attach-partition-local

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -1330,61 +1365,77 @@ bool StorageReplicatedMergeTree::executeLogEntry(LogEntry & entry)
+        {
+            Transaction transaction(*this);
+
+            if (renameTempPartAndAdd(part, nullptr, &transaction))
```
Maybe `renameTempPartAndReplace`?

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -1330,61 +1365,77 @@ bool StorageReplicatedMergeTree::executeLogEntry(LogEntry & entry)
+
+            if (renameTempPartAndAdd(part, nullptr, &transaction))
+                checkPartChecksumsAndCommit(transaction, part);
+
```
Maybe write something to `system.part_log`?

tavplubix replied:
I mean that `system.part_log` should contain `NewPart` event for every attached part, even if part was found locally (and was not fetched).

tavplubix commented:
Possibly we should change Changelog category to "Backward incompatible" and add a note that `ATTACH PART[ITION]` queries may not work during cluster upgrade.

tavplubix commented on tests/integration/test_attach_without_fetching/test.py:
```diff
@@ -0,0 +1,101 @@
+    with PartitionManager() as pm:
+        # The non-initiator replica downloads the data from initiator only.
+        # If something goes wrong and replica 1 wants to download data from replica 2, the test will fail.
+        pm.partition_instances(node_1, node_2)
```
```suggestion
        pm.partition_instances(node_1, node_2)
        pm.partition_instances(node_3, node_2)
```

myrrc replied:
Looks like this change would break the replica3 downloading data from replica2 and thus fail the test.

tavplubix replied:
No, `node_3` can download data from `node_1`. Yes, test will fail, but for different reason (because code is incorrect and replicas always execute `ATTACH_PART` as `GET_PART`).

myrrc replied:
Ok, as the replica can download the data not from the initiator only, but from every replica having the part, this change looks logical.

myrrc commented on docs/en/sql-reference/statements/alter/partition.md:
````diff
@@ -61,6 +61,8 @@ ALTER TABLE mt DROP PARTITION '2020-11-21';
 ALTER TABLE mt DROP PART 'all_4_4_0';
 ```
 
+Note: the command does NOT throw an exception on an invalid `partition_expr`.
````
The better way to write it would be "if the specified partition_expr does not exist", will correct.

tavplubix replied:
It's about DROP **PART**, not about partition

myrrc commented on docs/en/sql-reference/statements/alter/partition.md:
```diff
@@ -70,6 +72,8 @@ ALTER TABLE table_name DROP DETACHED PARTITION|PART partition_expr
 Removes the specified part or all parts of the specified partition from `detached`.
 Read more about setting the partition expression in a section [How to specify the partition expression](#alter-how-to-specify-part-expr).
 
+Note: the command does NOT throw an exception on an invalid `partition_expr`.
```
Will change as above.

myrrc commented on docs/en/sql-reference/statements/alter/partition.md:
```diff
@@ -85,7 +89,11 @@ ALTER TABLE visits ATTACH PART 201901_2_2_0;
+
+If the non-initiator replica, receiving the attach command, finds the part with the correct checksums in its own `detached` folder, 
+it attaches the data without fetching it from other replicas.
+If there is no part with the correct checksums, the data is downloaded from the replica-initiator.
```
I didn't change the docs in that part, you can see the diff.
I only added the change that the replica can attach the local data if found.

tavplubix replied:
It was incorrect, let's change it

myrrc commented on docs/en/sql-reference/statements/alter/partition.md:
````diff
@@ -95,7 +103,7 @@ So you can put data to the `detached` directory on one replica, and use the `ALT
 ```
 
-This query copies the data partition from the `table1` to `table2` adds data to exsisting in the `table2`. Note that data won’t be deleted from `table1`.
+This query copies the data partition from the `table1` to `table2`. Note that data won’t be deleted from `table1`.
````
This sentence part was a bit pointless, maybe it meant "Adding the data to the table2?"
If so, it's quite obvious that data is inserted in the table2.

tavplubix replied:
We also have `REPLACE PARTITION` query, which removes all previous data from the partition, so it's not obvious.
And we have `MOVE PARTITION`, but I don't remember what exactly it does.

myrrc replied:
And there is the sentence "Note that data won’t be deleted from `table1`", so it's perfectly correct.

tavplubix replied:
And also won't be deleted from `table2`

myrrc replied:
Ok, so I'll change it to "Note that data won't be deleted neither from table1 nor from table2".

myrrc commented on docs/en/sql-reference/statements/system.md:
````diff
@@ -264,6 +264,8 @@ Wait until a `ReplicatedMergeTree` table will be synced with other replicas in a
 SYSTEM SYNC REPLICA [db.]replicated_merge_tree_family_table_name
 ```
 
+Note that after running this statement the `[db.]replicated_merge_tree_family_table_name` fetches commands from other replicas.
````
The better way to write it would be "fetched command from other replicas' replicated logs to its own replicatino queue", will fix.

myrrc commented on src/Parsers/New/ClickHouseLexer.h:
```diff
@@ -13,51 +13,51 @@ namespace DB {
-    PERCENT = 206, PLUS = 207, QUERY = 208, QUOTE_DOUBLE = 209, QUOTE_SINGLE = 210, 
-    RBRACE = 211, RBRACKET = 212, RPAREN = 213, SEMICOLON = 214, SLASH = 215, 
-    UNDERSCORE = 216, MULTI_LINE_COMMENT = 217, SINGLE_LINE_COMMENT = 218, 
+    ADD = 1, AFTER = 2, ALIAS = 3, ALL = 4, ALTER = 5, AND = 6, ANTI = 7,
```
IDK

myrrc commented on src/Storages/StorageReplicatedMergeTree.h:
```diff
@@ -65,6 +66,8 @@ namespace DB
   * - if the part is corrupt (removePartAndEnqueueFetch) or absent during the check (at start - checkParts, while running - searchForMissingPart),
   *   actions are put on GET from other replicas;
   *
+  * TODO Update the GET part after rewriting the code (search locally).
```
Sorry, forgot to remove the place which was temporarily marked before.

myrrc commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -1330,61 +1365,77 @@ bool StorageReplicatedMergeTree::executeLogEntry(LogEntry & entry)
-    if (entry.type == LogEntry::GET_PART ||
-        entry.type == LogEntry::MERGE_PARTS ||
-        entry.type == LogEntry::MUTATE_PART)
+    if (entry.type == LogEntry::ATTACH_PART)
```
The check for the covering part is already made in the renameTempPartAndAdd, so no need to double-check it.

tavplubix replied:
It's a bit different check, `renameTempPartAndAdd` checks for `covered` parts, not for `covering` part. And `renameTempPartAndAdd` throws `LOGICAL_ERROR`.

myrrc commented:
> Possibly we should change Changelog category to "Backward incompatible" and add a note that ATTACH PART[ITION] queries may not work during cluster upgrade.

@tavplubix The proposed solution doesn't change the way the replicas process the commands if the data in detached/ is missing, I don't see the reason why it may be backward incompatible in any way.

myrrc committed:
Small fixes regarding the review

tavplubix commented:
> > Possibly we should change Changelog category to "Backward incompatible" and add a note that ATTACH PART[ITION] queries may not work during cluster upgrade.
> 
> @tavplubix The proposed solution doesn't change the way the replicas process the commands if the data in detached/ is missing, I don't see the reason why it may be backward incompatible in any way.

Previous versions of ClickHouse do not know anything about `ATTACH_PART` log entry. If someone run `ALTER ... ATTACH` query during cluster upgrade (when different version of ClickHouse are running on replicas), then replica with older version will fail to parse the entry from replication log. We should add a note about compatibility in changelog, so users will pay attention. Another option is to introduce a setting which disables the optimization for `ALTER ... ATTACH` (and disable it by default).

tavplubix commented:
Also it's not possible to rollback to older ClickHouse version after executing `ALTER ... ATTACH` query in new version.

myrrc commented:
So we decided to leave this ticket (and the corresponding RESTORE REPLICA) as a backward incompatible change, I'll update the description and the docs shortly.

tavplubix commented on docs/en/sql-reference/statements/alter/partition.md:
````diff
@@ -61,6 +61,9 @@ ALTER TABLE mt DROP PARTITION '2020-11-21';
 ALTER TABLE mt DROP PART 'all_4_4_0';
 ```
 
+Note: the command does NOT throw an exception if the specified part does not exist, 
````
https://github.com/ClickHouse/ClickHouse/blob/06678d650d228831425211bd45324b876e667ec0/src/Storages/StorageReplicatedMergeTree.cpp#L6203-L6204

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/attach-partition-local

myrrc committed:
Multiple small code and test updates

- Updated the docs to make everything clear.
- Multiple small logger fixes.
- Changed the attach_part command -- now it's after check for the
covering parts -- motivation is to do less work with the checksums
fetching.
- Better logging in the integration test.

myrrc committed:
Replaced the iteration algorithm in part finder

Now, instead of iterating through the directories, we iterate though
directories of on of the table disks (which doesn't give us a
substantial boost but is a bit neater to read).

- Updated the system.replication_queue command types.
- Fixed the part ptr being empty (added the checksum loading and
initialization).
- Removed extra logging.

myrrc committed:
Trying to investigate the errors after attaching the part.

+ A new entry is added to the part_log.

myrrc committed:
Investigating, why the checksums may match when they shouldn't.

myrrc commented:
Ok, as for now it seems the test started working (not failing due to the timeout), but here we have two issues:

1. The part with corrupted data may produce the same checksum as the part with the untouched data:
While running test we have the following logs for `replica1`:

```
> File data.bin, file size 67, file hash (14094526192655392907, 3703759728968193410)
Checksums files: 6, clean checksum: 173B97C0F462CA30629F9AA29CD4D72B
```

which represent the untouched data, and, after corrupting it on replica2, we have the following logs


```
Trying to attach part 1_2_2_0, checksum 173B97C0F462CA30629F9AA29CD4D72B
...
> File data.bin, file size 67, file hash (14094526192655392907, 3703759728968193410)
...
2021.03.17 20:53:00.875987 [ 61 ] {} <Trace> default.test (00a7fefe-b5a2-490f-83da-d6330d3d47ba): Checksums files: 6, path: detached/1_0_0_0, part checksum 173B97C0F462CA30629F9AA29C
D4D72B
```

So, the part with the corrupted data.bin file and the part with the untouched data produce the same checksums,
which obviously can't be true.
I tried explicitly deleting the .bin file on replica2, no luck.

2. The data part, being attached, correctly outputs in queries like SELECT * FROM, but in the logs there are strange messages like "part ... does not exist" even though it, well, exists. I think it's related to DataPartsExchanger somehow.

tavplubix commented:
> The part with corrupted data may produce the same checksum as the part with the untouched data
> So, the part with the corrupted data.bin file and the part with the untouched data produce the same checksums,
which obviously can't be true.
> I tried explicitly deleting the .bin file on replica2, no luck.

Just take a look at the code which loads checksums:
https://github.com/ClickHouse/ClickHouse/blob/3c11d444940e8b7c59d93a24366c9b2fb096ec2c/src/Storages/MergeTree/IMergeTreeDataPart.h#L136-L137

You need to either remove `checksums.txt` file or wait for `ReplicatedMergeTreePartCheckThread` to understand that attached part is corrupted. It will remove corrupted part and schedule fetching. The second option is better, as for me. But you may write tests for both options if you want.

But looks like test fails for different reason. Take a look at the server logs:
```
2021.03.17 23:18:26.210604 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Debug> executeQuery: (from 172.18.0.1:57212, using production parser) SELECT * FROM test ORDER BY n
2021.03.17 23:18:26.211535 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> ContextAccess (default): Access granted: SELECT(n) ON default.test
2021.03.17 23:18:26.212061 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Debug> default.test (b3562728-29ce-4a3e-9b7e-265c4152b5e1) (SelectExecutor): Key condition: unknown
2021.03.17 23:18:26.212099 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Debug> default.test (b3562728-29ce-4a3e-9b7e-265c4152b5e1) (SelectExecutor): MinMax index condition: unknown
2021.03.17 23:18:26.212977 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Debug> default.test (b3562728-29ce-4a3e-9b7e-265c4152b5e1) (SelectExecutor): Selected 10/10 parts by partition key, 10 parts by primary key, 10/20 marks by primary key, 10 marks to read from 10 ranges
2021.03.17 23:18:26.213191 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 9_0_0_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.213328 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 8_0_0_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.213423 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 7_0_0_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.213516 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 6_0_0_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.213622 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 5_0_0_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.213718 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 4_0_0_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.213818 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 3_0_0_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.213908 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 2_0_0_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.213998 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 1_2_2_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.214094 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> MergeTreeSelectProcessor: Reading 1 ranges from part 1_2_2_0, approx. 10 rows starting from 0
2021.03.17 23:18:26.214260 [ 9 ] {0b09da25-ff78-4df6-895c-5086943b6915} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
```

> The data part, being attached, correctly outputs in queries like SELECT * FROM

It does not.

There are two parts with the same name: `1_2_2_0`. And as you can see from the test logs some rows are duplicated and that's why the test failed:
```
E               @@ -1,3 +1,4 @@
E               +1
E                1
E                2
E                3
E               @@ -8,6 +9,7 @@
E                8
E                9
E                11
E               +11
E                12
E                13
... [and so on]
```

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -1778,21 +1849,20 @@ bool StorageReplicatedMergeTree::executeFetch(LogEntry & entry)
-        throw Exception("Too many total fetches from replicas, maximum: " + storage_settings_ptr->replicated_max_parallel_fetches.toString(),
+    if (storage_settings_ptr->replicated_max_parallel_fetches &&
+        total_fetches >= storage_settings_ptr->replicated_max_parallel_fetches)
+        throw Exception("Too many total fetches from replicas, maximum: " +
```
If you are here and changing code formatting, could you please also replace
`throw Exception("Some msg " + agr1 + " lorem ipsum " + arg2, ERROR_CODE)`
with
`throw Exception(ERROR_CODE, "Some msg {} lorem ipsum {}", arg1, arg2)`
?

tavplubix commented on src/Storages/MergeTree/MergeTreeDataPartChecksum.cpp:
```diff
@@ -293,11 +293,8 @@ String MergeTreeDataPartChecksums::getTotalChecksumHex() const
     SipHash hash_of_all_files;
 
-    for (const auto & elem : files)
+    for (auto && [name, checksum] : files)
```
?

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -1369,61 +1413,87 @@ bool StorageReplicatedMergeTree::executeLogEntry(LogEntry & entry)
+            checkPartChecksumsAndCommit(transaction, part);
+
+            writePartLog(PartLogElement::Type::NEW_PART, {},
+                0, // well, not really, but don't have the idea how to measure the time here TODO
```
0 is ok

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -1369,61 +1413,87 @@ bool StorageReplicatedMergeTree::executeLogEntry(LogEntry & entry)
+            writePartLog(PartLogElement::Type::NEW_PART, {},
+                0, // well, not really, but don't have the idea how to measure the time here TODO
+                part->name, part,
+                {part}, // not sure whether the initial parts vector should be empty or contain the part itself TODO
```
It should be empty

myrrc committed:
Added the solution to handle the corruption case

When the part data (e.g. data.bin) is corrupted, but the checksums.txt
is present -- explicitly deleting the checksums.txt.

Removed the extra logging, changes some exceptions message.

myrrc committed:
Added another test case to handle missing part data

myrrc committed:
Replaced removing checksums.txt with version

with the explicit data loading and verification.

If the function fails, the exception will re-throw upper,
cancelling the fetch in the handling of the replicated log entry,
but this event will also wake the PartCheckThread, which will
issue a re-fetch.

myrrc commented:
The bug with the double attach is due to our fetch in `executeFetch` and the fetch from the .`..PartCheckThread`.

myrrc commented:
So currently the main issue is:

1) The part (0_0_0_0) is attached locally on replica 2, which is confirmed by the test.

```
2021.03.22 21:38:48.959533 [ 23 ] {} <Warning> default.test (ReplicatedMergeTreePartCheckThread): Checking part 0_1_1_0
2021.03.22 21:38:48.961298 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 0_1_1_0, state: 2
2021.03.22 21:38:48.961411 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 1_2_2_0, state: 2
2021.03.22 21:38:48.961469 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 2_2_2_0, state: 2
2021.03.22 21:38:48.961517 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 3_0_0_0, state: 2
2021.03.22 21:38:48.961564 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 4_0_0_0, state: 2
2021.03.22 21:38:48.961611 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 5_0_0_0, state: 2
2021.03.22 21:38:48.961657 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 6_0_0_0, state: 2
2021.03.22 21:38:48.961704 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 7_0_0_0, state: 2
2021.03.22 21:38:48.961751 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 8_0_0_0, state: 2
2021.03.22 21:38:48.961797 [ 23 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Local name: 9_0_0_0, state: 2
```

2) Some magic happens (e.g. some MergeTree state in memory is corrupted)
3) PartCheckThread tries to find part locally but fails.

```
1519 2021.03.22 21:54:17.448089 [ 27 ] {} <Warning> default.test (ReplicatedMergeTreePartCheckThread): Checking part 0_1_1_0
1520 2021.03.22 21:54:17.449948 [ 27 ] {} <Trace> default.test (ReplicatedMergeTreePartCheckThread): Part 0_1_1_0 in zookeeper: true, locally: false
```

It deletes the data from the ZooKeeper.

4) Repica2 now thinks it doesn't have the part, so it trues to download the data from replica1 (which is in another partition), spamming the messages like "no part ... in table"
5) Replica 3 wants to download the data, but replica 2 deleted the info, so replica 2 tries to download data from replica 1 and fails on timeout each time.

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/attach-partition-local

myrrc committed:
Removed needless logging, fixed data part initialization while checking

in the detached folder

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/attach-partition-local

myrrc commented:
The bug with invalid MergeTree state was fixed by replacing `part_info` (in initializing part from detached/) to `actual_part_info`.

The part being attached from local data contained the old part information (which was outdated by the attach time), so the in-memory parts state was corrupted.

```
test_4ntuw6.replica1 (9ea5fc9d-0f71-408d-909b-cd5b7cda3504): Target part: name all_3_3_0, pid all, min 3, max 3, lev 0, mut 0
test_4ntuw6.replica1 (9ea5fc9d-0f71-408d-909b-cd5b7cda3504): Possible part: name all_0_0_0, pid all, min 0, max 0, lev 0 mut 0
test_4ntuw6.replica1 (9ea5fc9d-0f71-408d-909b-cd5b7cda3504): Possible part: name all_1_1_0, pid all, min 1, max 1, lev 0 mut 0
test_4ntuw6.replica1 (9ea5fc9d-0f71-408d-909b-cd5b7cda3504): Possible part: name all_2_2_0, pid all, min 2, max 2, lev 0 mut 0
executeQuery: Code: 232, e.displayText() = DB::Exception: No part all_3_3_0 in committed state
```

tavplubix committed:
Update MergeTreeDataPartChecksum.cpp

alexey-milovidov commented:
@tavplubix @myrrc Sorry, the test is flaky: https://clickhouse-test-reports.s3.yandex.net/22528/5ba6c7b7314f25140320027f81654610c71df027/integration_tests_(release).html#fail1

Revert?

myrrc commented:
Flacky test fixed in https://github.com/ClickHouse/ClickHouse/pull/22543.

gyuton commented:
Internal documentation ticket: DOCSUP-8233.
